File: backend/app/core/circuit_breaker.py
"""
Database Circuit Breaker

Protects PostgreSQL from cascading failures and connection pool exhaustion.

Features:
- Circuit breaker pattern (Open/Half-Open/Closed states)
- Query timeout enforcement
- Connection pool monitoring
- Automatic failover to cached data
- Exponential backoff for recovery
"""

import logging
import asyncio
from typing import Optional, Callable, Any, TypeVar, Dict
from datetime import datetime, timedelta
from enum import Enum
from functools import wraps
import time

logger = logging.getLogger(__name__)

T = TypeVar('T')


class CircuitState(str, Enum):
    """Circuit breaker states"""
    CLOSED = "closed"          # Normal operation
    OPEN = "open"              # Blocking requests (too many failures)
    HALF_OPEN = "half_open"    # Testing if service recovered


class CircuitBreakerOpenError(Exception):
    """Raised when circuit breaker is open"""
    pass


class DatabaseCircuitBreaker:
    """
    Circuit breaker for database operations.

    States:
    - CLOSED: Normal operation, requests pass through
    - OPEN: Too many failures, requests blocked (return cached data or error)
    - HALF_OPEN: Testing recovery, limited requests allowed

    Thresholds:
    - failure_threshold: Number of failures before opening circuit (default: 5)
    - recovery_timeout: Seconds to wait before trying half-open (default: 60)
    - success_threshold: Successes needed in half-open to close (default: 2)
    """

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: int = 60,
        success_threshold: int = 2,
        timeout_seconds: float = 30.0
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.success_threshold = success_threshold
        self.timeout_seconds = timeout_seconds

        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[datetime] = None
        self.last_state_change: datetime = datetime.utcnow()

        # Metrics
        self.total_calls = 0
        self.total_failures = 0
        self.total_successes = 0
        self.total_timeouts = 0
        self.total_circuit_open_rejections = 0

    async def call(
        self,
        func: Callable[..., Any],
        *args,
        fallback: Optional[Callable[..., Any]] = None,
        **kwargs
    ) -> Any:
        """
        Execute function with circuit breaker protection.

        Args:
            func: Async function to execute
            *args: Positional arguments for func
            fallback: Optional fallback function if circuit is open
            **kwargs: Keyword arguments for func

        Returns:
            Result from func or fallback

        Raises:
            CircuitBreakerOpenError: If circuit is open and no fallback provided
            asyncio.TimeoutError: If function exceeds timeout
        """
        self.total_calls += 1

        # Check if circuit should transition to half-open
        if self.state == CircuitState.OPEN:
            if self._should_attempt_reset():
                logger.info(f"Circuit breaker '{self.name}': Transitioning to HALF_OPEN")
                self.state = CircuitState.HALF_OPEN
                self.success_count = 0
                self.last_state_change = datetime.utcnow()
            else:
                # Circuit is open, use fallback or raise error
                self.total_circuit_open_rejections += 1
                logger.warning(
                    f"Circuit breaker '{self.name}': OPEN - Request blocked "
                    f"(failures: {self.failure_count}/{self.failure_threshold})"
                )

                if fallback:
                    logger.info(f"Circuit breaker '{self.name}': Using fallback function")
                    return await fallback(*args, **kwargs) if asyncio.iscoroutinefunction(fallback) else fallback(*args, **kwargs)
                else:
                    raise CircuitBreakerOpenError(
                        f"Circuit breaker '{self.name}' is OPEN. Database may be overloaded."
                    )

        # Execute function with timeout
        try:
            result = await asyncio.wait_for(
                func(*args, **kwargs),
                timeout=self.timeout_seconds
            )

            # Success
            self._on_success()
            return result

        except asyncio.TimeoutError:
            self.total_timeouts += 1
            logger.error(
                f"Circuit breaker '{self.name}': Query timeout after {self.timeout_seconds}s"
            )
            self._on_failure()

            if fallback:
                logger.info(f"Circuit breaker '{self.name}': Using fallback after timeout")
                return await fallback(*args, **kwargs) if asyncio.iscoroutinefunction(fallback) else fallback(*args, **kwargs)
            raise

        except Exception as e:
            logger.error(f"Circuit breaker '{self.name}': Error - {e}", exc_info=True)
            self._on_failure()

            if fallback:
                logger.info(f"Circuit breaker '{self.name}': Using fallback after error")
                return await fallback(*args, **kwargs) if asyncio.iscoroutinefunction(fallback) else fallback(*args, **kwargs)
            raise

    def _on_success(self):
        """Handle successful execution."""
        self.total_successes += 1
        self.failure_count = 0  # Reset failure count on success

        if self.state == CircuitState.HALF_OPEN:
            self.success_count += 1
            logger.info(
                f"Circuit breaker '{self.name}': HALF_OPEN success "
                f"({self.success_count}/{self.success_threshold})"
            )

            if self.success_count >= self.success_threshold:
                # Close the circuit
                logger.info(f"Circuit breaker '{self.name}': Transitioning to CLOSED")
                self.state = CircuitState.CLOSED
                self.success_count = 0
                self.failure_count = 0
                self.last_state_change = datetime.utcnow()

    def _on_failure(self):
        """Handle failed execution."""
        self.total_failures += 1
        self.failure_count += 1
        self.last_failure_time = datetime.utcnow()

        if self.state == CircuitState.HALF_OPEN:
            # Failure in half-open state, go back to open
            logger.warning(f"Circuit breaker '{self.name}': HALF_OPEN failed, returning to OPEN")
            self.state = CircuitState.OPEN
            self.success_count = 0
            self.last_state_change = datetime.utcnow()

        elif self.failure_count >= self.failure_threshold:
            # Too many failures, open the circuit
            logger.error(
                f"Circuit breaker '{self.name}': Threshold reached "
                f"({self.failure_count}/{self.failure_threshold}), transitioning to OPEN"
            )
            self.state = CircuitState.OPEN
            self.last_state_change = datetime.utcnow()

    def _should_attempt_reset(self) -> bool:
        """Check if enough time has passed to attempt reset."""
        if not self.last_failure_time:
            return True

        elapsed = (datetime.utcnow() - self.last_failure_time).total_seconds()
        return elapsed >= self.recovery_timeout

    def get_state(self) -> Dict[str, Any]:
        """Get current circuit breaker state and metrics."""
        return {
            "name": self.name,
            "state": self.state,
            "failure_count": self.failure_count,
            "success_count": self.success_count,
            "last_failure_time": self.last_failure_time.isoformat() if self.last_failure_time else None,
            "last_state_change": self.last_state_change.isoformat(),
            "config": {
                "failure_threshold": self.failure_threshold,
                "recovery_timeout": self.recovery_timeout,
                "success_threshold": self.success_threshold,
                "timeout_seconds": self.timeout_seconds
            },
            "metrics": {
                "total_calls": self.total_calls,
                "total_successes": self.total_successes,
                "total_failures": self.total_failures,
                "total_timeouts": self.total_timeouts,
                "total_circuit_open_rejections": self.total_circuit_open_rejections,
                "success_rate": (self.total_successes / self.total_calls * 100) if self.total_calls > 0 else 0,
                "failure_rate": (self.total_failures / self.total_calls * 100) if self.total_calls > 0 else 0
            }
        }

    def reset(self):
        """Manually reset circuit breaker to closed state."""
        logger.info(f"Circuit breaker '{self.name}': Manual reset to CLOSED")
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_state_change = datetime.utcnow()


# Global circuit breakers
_circuit_breakers: Dict[str, DatabaseCircuitBreaker] = {}


def get_circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: int = 60,
    success_threshold: int = 2,
    timeout_seconds: float = 30.0
) -> DatabaseCircuitBreaker:
    """
    Get or create a circuit breaker instance.

    Args:
        name: Unique name for this circuit breaker
        failure_threshold: Number of failures before opening circuit
        recovery_timeout: Seconds to wait before attempting recovery
        success_threshold: Successes needed to close circuit from half-open
        timeout_seconds: Query timeout in seconds

    Returns:
        DatabaseCircuitBreaker instance
    """
    if name not in _circuit_breakers:
        _circuit_breakers[name] = DatabaseCircuitBreaker(
            name=name,
            failure_threshold=failure_threshold,
            recovery_timeout=recovery_timeout,
            success_threshold=success_threshold,
            timeout_seconds=timeout_seconds
        )

    return _circuit_breakers[name]


def get_all_circuit_breakers() -> Dict[str, DatabaseCircuitBreaker]:
    """Get all registered circuit breakers."""
    return _circuit_breakers


# Decorator for automatic circuit breaker protection
def with_circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: int = 60,
    success_threshold: int = 2,
    timeout_seconds: float = 30.0,
    fallback: Optional[Callable] = None
):
    """
    Decorator to protect async functions with circuit breaker.

    Usage:
        @with_circuit_breaker(name="user_query", timeout_seconds=10.0)
        async def get_user(db: AsyncSession, user_id: str):
            result = await db.execute(select(User).where(User.id == user_id))
            return result.scalar_one_or_none()

    Args:
        name: Circuit breaker name
        failure_threshold: Failures before opening
        recovery_timeout: Recovery wait time
        success_threshold: Successes to close from half-open
        timeout_seconds: Query timeout
        fallback: Optional fallback function
    """
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        async def wrapper(*args, **kwargs):
            breaker = get_circuit_breaker(
                name=name,
                failure_threshold=failure_threshold,
                recovery_timeout=recovery_timeout,
                success_threshold=success_threshold,
                timeout_seconds=timeout_seconds
            )

            return await breaker.call(func, *args, fallback=fallback, **kwargs)

        return wrapper
    return decorator

File: frontend/src/utils/circuitBreaker.ts
/**
 * Circuit Breaker Pattern for Frontend
 * Prevents cascade failures when backend is unavailable
 */

type CircuitState = 'CLOSED' | 'OPEN' | 'HALF_OPEN';

interface CircuitBreakerOptions {
  failureThreshold?: number;      // Number of failures before opening
  resetTimeout?: number;          // Time in ms to wait before half-open
  halfOpenMaxCalls?: number;      // Max calls allowed in half-open state
  onStateChange?: (newState: CircuitState, oldState: CircuitState) => void;
}

interface CircuitMetrics {
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  rejectedCalls: number;
  lastFailureTime: number | null;
  consecutiveFailures: number;
}

export class CircuitBreaker {
  private state: CircuitState = 'CLOSED';
  private failureCount: number = 0;
  private lastFailureTime: number = 0;
  private halfOpenCalls: number = 0;
  private metrics: CircuitMetrics;

  private readonly failureThreshold: number;
  private readonly resetTimeout: number;
  private readonly halfOpenMaxCalls: number;
  private readonly onStateChange?: (newState: CircuitState, oldState: CircuitState) => void;

  constructor(options: CircuitBreakerOptions = {}) {
    this.failureThreshold = options.failureThreshold ?? 5;
    this.resetTimeout = options.resetTimeout ?? 30000; // 30 seconds
    this.halfOpenMaxCalls = options.halfOpenMaxCalls ?? 3;
    this.onStateChange = options.onStateChange;

    this.metrics = {
      totalCalls: 0,
      successfulCalls: 0,
      failedCalls: 0,
      rejectedCalls: 0,
      lastFailureTime: null,
      consecutiveFailures: 0,
    };
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    this.metrics.totalCalls++;

    // Check if circuit should transition from OPEN to HALF_OPEN
    if (this.state === 'OPEN') {
      if (Date.now() - this.lastFailureTime >= this.resetTimeout) {
        this.transitionTo('HALF_OPEN');
      } else {
        this.metrics.rejectedCalls++;
        throw new CircuitBreakerError('Circuit breaker is OPEN');
      }
    }

    // Check half-open state limits
    if (this.state === 'HALF_OPEN') {
      if (this.halfOpenCalls >= this.halfOpenMaxCalls) {
        this.metrics.rejectedCalls++;
        throw new CircuitBreakerError('Circuit breaker is HALF_OPEN and max calls reached');
      }
      this.halfOpenCalls++;
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  /**
   * Check if circuit allows requests
   */
  canExecute(): boolean {
    if (this.state === 'CLOSED') {
      return true;
    }

    if (this.state === 'OPEN') {
      if (Date.now() - this.lastFailureTime >= this.resetTimeout) {
        this.transitionTo('HALF_OPEN');
        return true;
      }
      return false;
    }

    // HALF_OPEN
    return this.halfOpenCalls < this.halfOpenMaxCalls;
  }

  /**
   * Get current circuit state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Get circuit metrics
   */
  getMetrics(): CircuitMetrics {
    return { ...this.metrics };
  }

  /**
   * Manually reset circuit to CLOSED
   */
  reset(): void {
    this.transitionTo('CLOSED');
    this.failureCount = 0;
    this.halfOpenCalls = 0;
    this.metrics.consecutiveFailures = 0;
  }

  private onSuccess(): void {
    this.metrics.successfulCalls++;
    this.metrics.consecutiveFailures = 0;

    if (this.state === 'HALF_OPEN') {
      // Reset to closed after successful call in half-open
      this.transitionTo('CLOSED');
      this.failureCount = 0;
      this.halfOpenCalls = 0;
    } else if (this.state === 'CLOSED') {
      // Reset failure count on success
      this.failureCount = 0;
    }
  }

  private onFailure(): void {
    this.metrics.failedCalls++;
    this.metrics.consecutiveFailures++;
    this.metrics.lastFailureTime = Date.now();
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (this.state === 'HALF_OPEN') {
      // Immediately open on failure in half-open
      this.transitionTo('OPEN');
      this.halfOpenCalls = 0;
    } else if (this.state === 'CLOSED' && this.failureCount >= this.failureThreshold) {
      // Open circuit when threshold reached
      this.transitionTo('OPEN');
    }
  }

  private transitionTo(newState: CircuitState): void {
    if (this.state !== newState) {
      const oldState = this.state;
      this.state = newState;

      console.log(`Circuit breaker: ${oldState} -> ${newState}`);

      if (this.onStateChange) {
        this.onStateChange(newState, oldState);
      }
    }
  }
}

/**
 * Custom error for circuit breaker rejections
 */
export class CircuitBreakerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CircuitBreakerError';
  }
}

/**
 * Global circuit breaker instances for different services
 */
const circuitBreakers: Map<string, CircuitBreaker> = new Map();

/**
 * Get or create a circuit breaker for a service
 */
export const getCircuitBreaker = (serviceName: string, options?: CircuitBreakerOptions): CircuitBreaker => {
  if (!circuitBreakers.has(serviceName)) {
    circuitBreakers.set(serviceName, new CircuitBreaker(options));
  }
  return circuitBreakers.get(serviceName)!;
};

/**
 * Reset all circuit breakers
 */
export const resetAllCircuitBreakers = (): void => {
  circuitBreakers.forEach(cb => cb.reset());
};

/**
 * Get status of all circuit breakers
 */
export const getAllCircuitBreakerStatus = (): Record<string, { state: CircuitState; metrics: CircuitMetrics }> => {
  const status: Record<string, { state: CircuitState; metrics: CircuitMetrics }> = {};
  circuitBreakers.forEach((cb, name) => {
    status[name] = {
      state: cb.getState(),
      metrics: cb.getMetrics(),
    };
  });
  return status;
};

export default CircuitBreaker;
